### ISSUE TITLE
[BUG] Remove res["pi"].detach().cpu()

### ISSUE THREAD

**Issue by pbarbarant** [alexisthual/fugw#42]
In ``fugw.mapping`` :
```
# Store variables of interest in model
self.pi = res["pi"].detach().cpu()
```

Causes problems when dealing with barycenters as the transport plans get detached before the next barycenter iteration. Discovered with @antoinecollas.

**Comment by alexisthual** [comment]
Sorry, I'm a bit late in the game here, but I think it's important to store the final object using `cpu`, otherwise you might not be able to load it on machines which do not have a gpu (I might be wrong but I recall something of that kind happening to me).

**Comment by pbarbarant** [comment]
> Sorry, I'm a bit late in the game here, but I think it's important to store the final object using `cpu`, otherwise you might not be able to load it on machines which do not have a gpu (I might be wrong but I recall something of that kind happening to me).

Wouldn't it be better to upgrade `save_mapping` / `load_mapping` utilities to by send the `mapping` object to the `cpu` by default before dumping it ?

**Comment by pbarbarant** [comment]
I'm afraid that by detaching to the cpu right after fitting, we might introduce a large memory overhead when computing barycenters.

**Comment by alexisthual** [comment]
> > Sorry, I'm a bit late in the game here, but I think it's important to store the final object using `cpu`, otherwise you might not be able to load it on machines which do not have a gpu (I might be wrong but I recall something of that kind happening to me).
> 
> Wouldn't it be better to upgrade `save_mapping` / `load_mapping` utilities to by send the `mapping` object to the `cpu` by default before dumping it ?

That sounds like a very good idea :)

**Comment by pbarbarant** [comment]
> > > Sorry, I'm a bit late in the game here, but I think it's important to store the final object using `cpu`, otherwise you might not be able to load it on machines which do not have a gpu (I might be wrong but I recall something of that kind happening to me).
> > 
> > 
> > Wouldn't it be better to upgrade `save_mapping` / `load_mapping` utilities to by send the `mapping` object to the `cpu` by default before dumping it ?
> 
> That sounds like a very good idea :)

Reopening this issue
